### ISSUE TITLE
Support the new Configuration Cache

### ISSUE THREAD

**Issue by Mahoney** [autonomousapps/dependency-analysis-gradle-plugin#306]
**Is your feature request related to a problem? Please describe.**
I'd like to enable the new configuration cache:

https://docs.gradle.org/6.7-rc-5/userguide/configuration_cache.html

but it has lots of problems with this plugin.

**Describe the solution you'd like**
It would be nice if there were no errors in this plugin when running `gradle --configuration-cache`

**Additional context**
Obviously this is just a nice to have, everything works without it; just thought I'd put it on the radar.

**Comment by autonomousapps** [comment]
I agree and would love to have this plugin support the configuration cache, but at the moment it's impossible. Some of the data the plugin needs to operate is not available in a conf-cache-friendly way. There are open issues on the Gradle issue tracker, but as far as I know, no eta on resolving them.

**Comment by ZacSweers** [comment]
Stacktrace for reference

```
org.gradle.api.InvalidUserCodeException: Listener registration 'TaskExecutionGraph.addTaskExecutionListener' by org.gradle.execution.taskgraph.DefaultTaskExecutionGraph@167d2e92 is unsupported.
	at org.gradle.configurationcache.initialization.DefaultConfigurationCacheProblemsListener.onBuildScopeListenerRegistration(ConfigurationCacheProblemsListener.kt:90)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.event.DefaultListenerManager$ListenerDetails.dispatch(DefaultListenerManager.java:398)
	at org.gradle.internal.event.DefaultListenerManager$ListenerDetails.dispatch(DefaultListenerManager.java:380)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:61)
	at org.gradle.internal.event.DefaultListenerManager$EventBroadcast$ListenerDispatch.dispatch(DefaultListenerManager.java:368)
	at org.gradle.internal.event.DefaultListenerManager$EventBroadcast.dispatch(DefaultListenerManager.java:179)
	at org.gradle.internal.event.DefaultListenerManager$EventBroadcast.dispatch(DefaultListenerManager.java:153)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy21.onBuildScopeListenerRegistration(Unknown Source)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph.notifyListenerRegistration(DefaultTaskExecutionGraph.java:299)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph.addTaskExecutionListener(DefaultTaskExecutionGraph.java:255)
	at com.autonomousapps.DependencyAnalysisPlugin.apply(DependencyAnalysisPlugin.kt:104)
	at com.autonomousapps.DependencyAnalysisPlugin.apply(DependencyAnalysisPlugin.kt:47)
```

@autonomousapps could you maybe link the upstream gradle issue?

**Comment by autonomousapps** [comment]
@ZacSweers which version of the plugin are you on? I haven't used a task execution listener for a while.

When I'm at my desk I'll link the issue I was thinking about, but I suspect your issue is different.

**Comment by ZacSweers** [comment]
0.61.0

**Comment by autonomousapps** [comment]
Try 0.70.0?

**Comment by autonomousapps** [comment]
For the record, this is the issue I filed against Gradle that prevents this plugin's execution from being fully conf-cache-ready. Please note that simply applying the plugin shouldn't break other aspects of the build: it's simply that the plugin's tasks will always require a full configuration until Gradle fixes the issue (which they have acknowledged they ought to do).

@ZacSweers it was v0.65.0 that [removed](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/blob/main/CHANGELOG.md#version-0650) that task execution listener.

**Comment by ZacSweers** [comment]
0.70 reports no issues for us 👍

**Comment by dsvoronin** [comment]
JFYI, version 0.77, got:

```
Configuration cache state could not be cached: field 'usedTransitiveComponents$delegate' from type 'com.autonomousapps.tasks.AdvicePerVariantTask': error writing value of type 'kotlin.UnsafeLazyImpl'
Configuration cache state could not be cached: field 'value' from type 'kotlin.InitializedLazyImpl': error writing value of type 'java.util.LinkedHashSet'
> Configuration cache state could not be cached: field 'dependency' from type 'com.autonomousapps.internal.TransitiveComponent': error writing value of type 'com.autonomousapps.advice.Dependency'
  > Configuration cache state could not be cached: field 'group$delegate' from type 'com.autonomousapps.advice.Dependency': error writing value of type 'kotlin.SynchronizedLazyImpl'
    > begin 0, end -1, length 10
```

**Comment by autonomousapps** [comment]
@dsvoronin that issue could probably be resolved quickly. It probably isn't really worth it for that property to be lazy. If you want to open a PR, I think it's just a one-line change.

In general though, this plugin cannot be CC-compatible until Gradle provides a way to examine the resolved dependency graph in a compliant way.

**Comment by eskatos** [comment]
gradle/gradle#18729 won't be in 7.4, it needs more work and is being reverted for now in gradle/gradle#19119.

**Comment by autonomousapps** [comment]
Thanks for letting me know, @eskatos!

**Comment by mvegter** [comment]
Found two other issues related to configuration cache

> Task `:XYZ:findDeclaredProcsMain` of type `com.autonomousapps.tasks.FindDeclaredProcsTask`: cannot serialize object of type 'com.autonomousapps.services.InMemoryCache', a subtype of 'org.gradle.api.services.BuildService', as these are not supported with the configuration cache.

> Task `:XYZ:graphViewMain` of type `com.autonomousapps.tasks.GraphViewTask`: cannot serialize object of type 'org.gradle.api.internal.artifacts.configurations.DefaultConfiguration', a subtype of 'org.gradle.api.artifacts.Configuration', as these are not supported with the configuration cache.

**Comment by eric-labelle** [comment]
FYI - Additional to @mvegter issues, I'm also seeing this one using `0.80.0` version and Gradle `7.4`:
> Task `:XYZ:locateDependenciesDebug` of type `com.autonomousapps.tasks.LocateDependenciesTask`: cannot serialize object of type 'org.gradle.api.internal.artifacts.configurations.DefaultConfigurationContainer', a subtype of 'org.gradle.api.artifacts.ConfigurationContainer', as these are not supported with the configuration cache.

**Comment by autonomousapps** [comment]
https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/pull/680 resolves https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/issues/306#issuecomment-1045985505

https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/issues/306#issuecomment-1048848550 is not an issue since 1.0.

**Comment by autonomousapps** [comment]
Will release [the partial support](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/pull/680) as 1.2.2.

**Comment by autonomousapps** [comment]
Ended up releasing as 1.3.0, since there was a new feature as well.

**Comment by ianbrandt** [comment]
Filed #885 for the issue described above (with reference to the possibly related gradle/gradle#24636).

**Comment by vicky17d** [comment]
Just tried on one of my projects with version 1.14.1 and the latest one at the moment i.e. 1.20.0
No issues with storing configuration cache but it still does end up discarding it every time
`
0 problems were found storing the configuration cache.

See the complete report at file:///Users/<username>/buildHealth/<project-name>/build/reports/configuration-cache/c9rypw15z23d3mxbpt39glbci/9vgbv9endq14ao2zoc6mlt1j9/configuration-cache-report.html

BUILD SUCCESSFUL in 11s
145 actionable tasks: 116 executed, 17 from cache, 12 up-to-date

Publishing build scan...

Configuration cache entry discarded.  // <-----------
`

Curious why it is being discarded even it says 0 problems found storing the cache? Something under the hood? 
Thanks

**Comment by TWiStErRob** [comment]
@vicky17d discard happens when a task calls [`Task.notCompatibleWithConfigurationCache(cause)`](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/pull/680/files#diff-f01700fb3f8012ad7473c590e4c483c3cb5dcb55d329a2ec866728960dce527bR34).

**Comment by gabrielittner** [comment]
Another configuration cache related thing with Gradle 8.1.1, there seems to be a lot of file system access at configuration time. For any project I see this in the config cache report: 
<img width="339" alt="Screenshot 2023-06-22 at 11 13 21" src="https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/assets/1358105/70a01686-a37f-41dc-9605-e482274a4c37">
It's not that bad but it adds up to thousands or tens of thousands of inputs that Gradle needs to keep track of in big projects

**Comment by NikolayMetchev** [comment]
Can somebody tell me if this is still blocked by gradle or is there a way forward?

**Comment by jjohannes** [comment]
I think this is not blocked anymore. Gradle now allows resolution results as task inputs. It's sometimes a bit tedious to figure out how to use it correctly, but it works.

This issue just™️ needs to be done now. 

(Hint: I think @autonomousapps would be happy to work on this if someone would sponsor the work 😉)

**Comment by NikolayMetchev** [comment]
Thanks @jjohannes Just curious if payment for one feature is an option and what would that cost?

**Comment by autonomousapps** [comment]
Hi @NikolayMetchev! I am definitely open to sponsorships to help me prioritize where to focus my efforts. There's only so much time in the day, and this plugin gets a lot of attention since it provides important functionality available nowhere else in the Gradle world (to my knowledge). When someone is willing to support development financially, that is a very strong signal about how important an issue is for them and probably others as well.

I see you are active in the Gradle Community Slack as well. We can talk more there if you like.

**Comment by bddckr** [comment]
I've been using this plugin with great success, even with the config cache turned on, for quite a while now. To ensure my CI pipeline is still fast, I've changed the pipeline to run all config cache-compatible tasks first. In a second Gradle invocation, it then runs the few remaining tasks that don't support the config cache (like those from this plugin).

However, I've started to notice that the config cache is invalidated by Gradle when I change any source file in any of my projects. That means I'm not using the config cache at all...

Thanks to the Gradle config cache reports, I was able to identify the issue: It's due to all source set roots being marked as input for the tasks of this plugin - just like the comment [above](https://github.com/autonomousapps/dependency-analysis-gradle-plugin/issues/306#issuecomment-1602299268) says.

So, while that comment calls out that this comes with a performance penalty because Gradle has to track all those folders as input, there's also the issue of this resulting in the config cache being invalidated whenever any of your sources change!

To fix this, I've changed my convention plugin that applies this plugin: I now only apply this plugin when `gradle.startParameter.isConfigurationCacheRequested` is `true`. That way, I can apply this plugin in my pipeline in the second Gradle invocation mentioned above by adding `--no-configuration-cache` to the arguments, giving me the advantages of the config cache for all other tasks and speeding up my local development again. The config cache is still properly used for all other tasks with this change.

---

tl;dr:
* If you apply this plugin, changing any source file will lead to invalidation of the config cache, making it mostly useless.
* If you split your Gradle invocations into using or not using the config cache, you can gain back the speed improvements the config cache gives you.
    * For that, consider conditionally applying this plugin - I use `gradle.startParameter.isConfigurationCacheRequested` to decide.

Hopefully, this can help others 😄

**Comment by TWiStErRob** [comment]
@bddckr I did a quick try and didn't notice the cache invalidation you mentioned:
 * I ran `gradlew :glide4:clean :capture_image:clean :capture_image:compileDebugATSo --no-build-cache` (which is just something I'm working on at the moment),
 * then changed a file in a folder which is mentioned in the CC report (in `:capture_image > src/main/java`)
 *  then re-ran the same command and got `Reusing configuration cache.`.

plugin 1.22.0, Gradle: 8.2.1, AGP: 8.1.1

**Comment by bddckr** [comment]
That made me find the issue now - thank you for the ping to make me notice that!

It turns out the issue is simply because my convention plugin (that applies this plugin) lives in another included build. I moved the convention plugin to the main Gradle build as part of applying the above "fix" I explained. If I now remove the conditional check and instead _always_ apply this plugin, it works fine again. The config cache is reused.

I've edited my comment above so as not to lead others astray.

**Comment by autonomousapps** [comment]
I believe this is now complete.

**Comment by NikolayMetchev** [comment]
> I believe this is now complete.
Thanks @autonomousapps This is fantastic.

It looks like 1.27.0 doesn't yet have the necessary changes. When can we expect a release with this fix in it?

**Comment by Mahoney** [comment]
`gradle --configuration-cache` exits with 0 and no errors for me with 1.27.0 applied.

**Comment by NikolayMetchev** [comment]
> `gradle --configuration-cache` exits with 0 and no errors for me with 1.27.0 applied.

I get a message at the end of the build that is along the lines of: 
```
Configuration cache entry discarded because incompatible tasks were found: '*:graphViewMain
```

If you run `buildHealth` twice does it actually use the configuration cache for you?

I am pretty sure this commit is needed which isn't in 1.27.0
https://github.com/autonomousapps/dependency-analysis-gradle-plugin/commit/424e0fbe701fc0b08fcafb1174c43546d92823f8

**Comment by Mahoney** [comment]
You're right, sorry, couldn't remember how to check it.

**Comment by autonomousapps** [comment]
You can use the latest snapshot. I'll cut a release soon.

**Comment by Mahoney** [comment]
Appears to be in 1.28.0 - ` gw --configuration-cache buildHealth` returns 0 with "Reusing configuration cache."
